pkg/services/object/get/ec.go

```diff
@@ -60,6 +62,129 @@ func (s *Service) copyECObject(ctx context.Context, cnr cid.ID, parent oid.ID, s
 	return nil
 }
 
+func (s *Service) copyECObjectHeader(ctx context.Context, dst internal.HeaderWriter, cnr cid.ID, parent oid.ID,
+	sTok *session.Object, ecRules []iec.Rule, sortedNodeLists [][]netmap.NodeInfo) error {
+	hdr, err := s.getECObjectHeader(ctx, cnr, parent, sTok, ecRules, sortedNodeLists)
+	if err != nil {
+		return err
+	}
+
+	if err := dst.WriteHeader(&hdr); err != nil {
+		return fmt.Errorf("write header: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Service) getECObjectHeader(ctx context.Context, cnr cid.ID, id oid.ID, sTok *session.Object,
+	ecRules []iec.Rule, sortedNodeLists [][]netmap.NodeInfo) (object.Object, error) {
+	localNodeKey, err := s.keyStore.GetKey(nil)
+	if err != nil {
+		return object.Object{}, fmt.Errorf("get local SN private key: %w", err)
+	}
+
+	// TODO: sort EC rules by complexity and try simpler ones first. Note that rule idxs passed as arguments must be kept.
+	//  https://github.com/nspcc-dev/neofs-node/issues/3563
+	// TODO: limit per-rule context? https://github.com/nspcc-dev/neofs-node/issues/3560
+	var firstErr error
+	for i := range ecRules {
+		hdr, err := s.getECObjectHeaderByRule(ctx, *localNodeKey, cnr, id, sTok, ecRules[i], sortedNodeLists[i])
+		if err == nil || errors.Is(err, apistatus.ErrObjectAlreadyRemoved) || errors.Is(err, ctx.Err()) {
+			return hdr, err
+		}
+
+		if i == 0 {
+			firstErr = err
+		}
+
+		s.log.Info("failed to fetch object header by EC rule",
+			zap.Stringer("container", cnr), zap.Stringer("object", id), zap.Stringer("rule", ecRules[i]),
+			zap.Int("rulesLeft", len(ecRules)-i-1), zap.Error(err))
+	}
+
+	return object.Object{}, fmt.Errorf("%w: failed processing of all %d EC rules, first error: %w",
+		apistatus.ErrObjectNotFound, len(ecRules), firstErr)
+}
+
+func (s *Service) getECObjectHeaderByRule(ctx context.Context, localNodeKey ecdsa.PrivateKey, cnr cid.ID, id oid.ID,
+	sTok *session.Object, rule iec.Rule, sortedNodes []netmap.NodeInfo) (object.Object, error) {
+	totalParts := int(rule.DataPartNum + rule.ParityPartNum)
+
+	var hdrTarget atomic.Value
+	var firstErrTarget atomic.Value
+
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(min(totalParts, 3))
+
+	for i := range sortedNodes {
+		node := sortedNodes[i]
+		eg.Go(func() error {
```

Comment:
Concurrent execution is an overkill here. Every part has the header, so chances of the first try to succeed are very high.

Reply:
reasonable, done